src/Logic.lidr

```diff
@@ -0,0 +1,1296 @@
+from the propositional to the boolean claim), we used a simple induction on
+\idr{k}. On the other hand, the converse (the \idr{evenb_double_conv} exercise)
+required a clever generalization, since we can't directly prove \idr{(k ** n =
+double k) -> evenb n = True}.
+
+For these examples, the propositional claims are more useful than their boolean
+counterparts, but this is not always the case. For instance, we cannot test
+whether a general proposition is true or not in a function definition; as a
+consequence, the following code fragment is rejected:
+
+```idris
+is_even_prime : Nat -> Bool
+is_even_prime n = if n = 2 then True else False
+```
+
+Idris complains that \idr{n = 2} has type \idr{Type}, while it expects an
+element of \idr{Bool} (or some other inductive type with two elements). The
+reason for this error message has to do with the _computational_ nature of
+Idris's core language, which is designed so that every function that it can
+express is computable and total. One reason for this is to allow the extraction
+of executable programs from Idris developments. As a consequence, \idr{Type} in
+Idris does _not_ have a universal case analysis operation telling whether any
+given proposition is true or Void, since such an operation would allow us to
+write non-computable functions.
+
+Although general non-computable properties cannot be phrased as boolean
+computations, it is worth noting that even many _computable_ properties are
+easier to express using \idr{Type} than \idr{Bool}, since recursive function
+definitions are subject to significant restrictions in Idris. For instance, the
+next chapter shows how to define the property that a regular expression matches
+a given string using \idr{Type}. Doing the same with \idr{Bool} would amount to
+writing a regular expression matcher, which would be more complicated, harder to
+understand, and harder to reason about.
+
+Conversely, an important side benefit of stating facts using booleans is
+enabling some proof automation through computation with Idris terms, a technique
+known as _proof by reflection_. Consider the following statement:
+
+> even_1000 : (k ** 1000 = double k)
+
+The most direct proof of this fact is to give the value of \idr{k} explicitly.
+
+> even_1000 = (500 ** Refl)
+
+On the other hand, the proof of the corresponding boolean statement is even
+simpler:
+
+> even_1000' : evenb 1000 = True
+> even_1000' = Refl
+
+What is interesting is that, since the two notions are equivalent, we can use
+the boolean formulation to prove the other one without mentioning the value 500
+explicitly:
+
+> even_1000'' : (k ** 1000 = double k)
+> even_1000'' = (fst $ even_bool_prop {n=1000}) Refl
+
+Although we haven't gained much in terms of proof size in this case, larger
+proofs can often be made considerably simpler by the use of reflection. As an
+extreme example, the Coq proof of the famous _4-color theorem_ uses reflection
```

Comment:
Should we add link such as the following? http://www.ams.org/journals/notices/200811/tx081101382p.pdf